data-collection-scripts/cdc-vaccinations-timeseries/make_timeseries.py

```diff
@@ -0,0 +1,67 @@
+import json
+import git
+import csv
+import sys
+import click
+from collections import defaultdict
+
+# we want our columns to come out in the desired order, and older pythons don't guarantee dict ordering
+MIN_PYTHON = (3, 7)
+if sys.version_info < MIN_PYTHON:
+    sys.exit("Python %s.%s or later is required.\n" % MIN_PYTHON)
+
+
+@click.command()
+@click.option('--strip-duplicate-days/--no-strip-duplicate-days', default=False,
+              help='Process the data to only output one set of data for each day')
+def main(strip_duplicate_days):
+    repo = git.Repo("../../")
+    path = "data/cdc_vaccinations.json"
+
+    # fetch all the git commits that updated the data file
+    # thanks to https://stackoverflow.com/q/28803626
+    revlist = (
+        (commit, (commit.tree / path).data_stream.read())
+        for commit in repo.iter_commits(paths=path)
+    )
+
+    # build up a dict of the data history by looking at each commit in the git history
+    data = {}
+    for commit, filecontents in revlist:
+        data[commit.committed_datetime] = json.loads(filecontents)
+
+    # go through the data and reformat it into one line per batch/state
+    # if --strip-duplicate-days is set, we only keep the latest set of data for any state/date pair
+    out_data = []
+    out_cols = {"runid": None}
+    seen_data = defaultdict(set)  # dict with date keys holding the states we've seen for that date
+    for data_batch in data:
+        data_batch = data[data_batch]
+        for state_data in data_batch["vaccination_data"]:
+            state_data["runid"] = data_batch["runid"]
+
+            # there's an anomaly in the data where one day has the date in the wrong format.
+            # reformat it, hoping this was a one-time glitch
+            if state_data["Date"] == "01/12/2021":
+                state_data["Date"] = "2021-01-12"
+
+            # have we seen this state/date before?
+            if strip_duplicate_days and state_data["ShortName"] in seen_data[state_data["Date"]]:
+                continue
+
+            # keep track of all the columns that exist as we go (we'll need that list for the CSV output)
+            for k in state_data.keys():
+                out_cols[k] = None
```

Comment:
if i'm not missing anything, the values in `out_cols` aren't used? so maybe a list or a set instead

Reply:
You're not missing anything; I'm using the keys of a dict as a set because Python sets don't preserve order and dicts are now guaranteed to do so, and if you use a set (without using some ordered set library off of PyPI or implementing my own), you end up with the CSV columns in random order, while a dict preserves the ordering. I'm open to a better way to do this.